lib/strategy/FragmentationStrategyShape.ts

```diff
@@ -0,0 +1,292 @@
+        if (positionContainer !== -1) {
+          const generate = this.shouldGenerate();
+          // Skip the resource based on the generation probability
+          if (!generate) {
+            this.irisHandled.add(iri);
+            this.resourcesHandled.add(resourceId);
+            return;
+          }
+          const podIRI = iri.slice(0, Math.max(0, positionContainer));
+
+          const shapeTreeIRI = `${podIRI}/${FragmentationStrategyShape.shapeTreeFileName}`;
+
+          const promises: Promise<void>[] = [];
+          // We add a triple in each file to locate the shape index if it is enable
+          if (this.tripleShapeTreeLocator === true) {
+            promises.push(FragmentationStrategyShape.generateShapeTreeLocator(quadSink, `${podIRI}/`, shapeTreeIRI, iri));
+          }
+
+          if (!this.resourcesHandled.has(resourceId)) {
+            promises.push(FragmentationStrategyShape.generateShapeIndexInformation(quadSink,
+              resourceId,
+              podIRI,
+              shapeTreeIRI,
+              directory,
+              shape,
+              positionContainerResourceNotInRoot === -1));
+            this.irisHandled.add(iri);
+            this.resourcesHandled.add(resourceId);
+            await Promise.all(promises);
+            return;
+          }
+          if (promises.length === 1) {
+            this.irisHandled.add(iri);
+            this.resourcesHandled.add(resourceId);
+            await promises[0];
+            return;
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Generate all the mandatory shape index information and push them into a quad sink
+   * @param {IQuadSink} quadSink - a quad sink
+   * @param {string} resourceId - the id of the resource, may be used to locate the target of the shape index
+   * @param {string} podIRI - the iri of the pod
+   * @param {string} shapeTreeIRI - the iri of the shapetree
+   * @param {string} directory - the folder bounded binded by the shape
+   * @param {string} shape - ShExC string
+   * @param {boolean} isInRootOfPod - indicate if the resource is at the root of the pod
+   */
+  public static async generateShapeIndexInformation(quadSink: IQuadSink,
+    resourceId: string,
+    podIRI: string,
+    shapeTreeIRI: string,
+    directory: string,
+    shape: string,
+    isInRootOfPod: boolean): Promise<void> {
+    const shapeIRI = `${podIRI}/${directory}_shape`;
+    // In Solid the path to a container end with a trailing "/"
+    // hence when the resource is not in the root the content iri must be different.
+    const contentIri = isInRootOfPod ? resourceId : `${podIRI}/${directory}/`;
+    const promises = [
+      FragmentationStrategyShape.generateShape(quadSink, shapeIRI, shape),
+      FragmentationStrategyShape.generateShapetreeTriples(quadSink, shapeTreeIRI, shapeIRI, isInRootOfPod, contentIri),
+    ];
+
+    await Promise.all(promises);
```

Comment:
Is there a reason to not just await each promise inline?

Reply:
I just wanted to exploit the concurrence instead of waiting